crates/ruff/src/rules/pylint/rules/redefined_loop_name.rs

```diff
@@ -0,0 +1,250 @@
+    ///     i = 9
+    ///     print(i)  # prints 9 every iteration
+    ///
+    /// for i in range(10):
+    ///     for i in range(10):  # original value overwritten
+    ///         pass
+    ///     print(i)  # also prints 9 every iteration
+    ///
+    /// with path1.open() as f:
+    ///     with path2.open() as f:
+    ///         f = path2.open()
+    ///     print(f.readline())  # prints a line from path2
+    /// ```
+    pub struct RedefinedLoopName {
+        pub name: String,
+        pub outer_kind: BindingKind,
+        pub inner_kind: BindingKind,
+    }
+);
+impl Violation for RedefinedLoopName {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let RedefinedLoopName {
+            name,
+            outer_kind,
+            inner_kind,
+        } = self;
+        format!(
+            "Outer {outer_kind} variable `{name}` overwritten by {inner_kind} target with same name"
+        )
+    }
+}
+
+struct InnerForWithAssignNamesVisitor<'a> {
+    locator: &'a Locator<'a>,
+    name_ranges: Vec<(Range, BindingKind)>,
+}
+
+impl<'a, 'b> Visitor<'b> for InnerForWithAssignNamesVisitor<'_>
+where
+    'b: 'a,
+{
+    fn visit_stmt(&mut self, stmt: &'b Stmt) {
+        // Collect target names.
+        match &stmt.node {
+            // For and async for.
+            StmtKind::For { target, .. } | StmtKind::AsyncFor { target, .. } => {
+                self.name_ranges.extend(name_ranges_from_expr(
+                    target,
+                    self.locator,
+                    BindingKind::For,
+                ));
+            }
+            // With.
+            StmtKind::With { items, .. } => {
+                self.name_ranges
+                    .extend(name_ranges_from_with_items(items, self.locator));
+            }
+            // Assignment, augmented assignment, and annotated assignment.
+            StmtKind::Assign { targets, .. } => {
+                self.name_ranges
+                    .extend(name_ranges_from_assign_targets(targets, self.locator));
+            }
+            StmtKind::AugAssign { target, .. } | StmtKind::AnnAssign { target, .. } => {
+                self.name_ranges.extend(name_ranges_from_expr(
+                    target,
+                    self.locator,
+                    BindingKind::Assignment,
+                ));
+            }
+            _ => {}
+        }
+        // Decide whether to recurse.
+        match &stmt.node {
+            // Don't recurse into blocks that create a new scope.
+            StmtKind::ClassDef { .. } => {}
+            StmtKind::FunctionDef { .. } => {}
+            // Otherwise, do recurse.
+            _ => {
+                visitor::walk_stmt(self, stmt);
+            }
+        }
+    }
+}
+
+fn name_ranges_from_expr<'a, U>(
+    target: &'a Expr<U>,
+    locator: &'a Locator,
+    kind: BindingKind,
+) -> impl Iterator<Item = (Range, BindingKind)> + 'a {
+    find_names(target, locator).map(move |item| (item, kind))
```

Comment:
I think it'd be better to avoid the lexer here, I'm guessing it's the cause of the false positives. Can we instead look for `ExprKind::Name` references with `ExprContext::Store`?

Reply:
Done. I'm comparing the target expressions structurally, while still respecting unpacking of tuples and lists, which resolves the false positives you mentioned.